File: tutorials/tour-of-restate-rust/src/checkout_service.rs
use crate::auxiliary::email_client::EmailClient;
use crate::auxiliary::payment_client::PaymentClient;
use restate_sdk::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CheckoutRequest {
    pub(crate) user_id: String,
    pub(crate) tickets: Vec<String>,
}

#[restate_sdk::service]
pub(crate) trait CheckoutService {
    async fn handle(request: Json<CheckoutRequest>) -> Result<bool, HandlerError>;
}

pub struct CheckoutServiceImpl;

impl CheckoutService for CheckoutServiceImpl {
    async fn handle(
        &self,
        mut ctx: Context<'_>,
        Json(CheckoutRequest { user_id, tickets }): Json<CheckoutRequest>,
    ) -> Result<bool, HandlerError> {
        let total_price = tickets.len() as f64 * 40.0;

        let idempotency_key = ctx.rand_uuid().to_string();

        let pay_client = PaymentClient::new();
        let success = ctx
            .run(|| pay_client.call(&idempotency_key, total_price))
            .await?;

        let email_client = EmailClient::new();
        if success {
            ctx.run(|| email_client.notify_user_of_payment_success(&user_id))
                .await?;
        } else {
            ctx.run(|| email_client.notify_user_of_payment_failure(&user_id))
                .await?;
        }
        Ok(success)
    }
}

File: python/tutorials/tour-of-restate-python/tour/part4/checkout_service.py
# Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
#
# This file is part of the Restate examples,
# which is released under the MIT license.
#
# You can find a copy of the license in the file LICENSE
# in the root directory of this repository or package or at
# https://github.com/restatedev/examples/
import uuid
from typing import TypedDict, List
from restate.context import ObjectContext, Serde
from restate.service import Service

from tour.auxiliary.email_client import EmailClient
from tour.auxiliary.payment_client import PaymentClient


class Order(TypedDict):
    user_id: str
    tickets: List[str]


payment_client = PaymentClient()
email_client = EmailClient()

checkout = Service("CheckoutService")


# <start_checkout>
@checkout.handler()
async def handle(ctx: ObjectContext, order: Order) -> bool:
    total_price = len(order["tickets"]) * 40

    idempotency_key = str(ctx.uuid())

    async def pay():
        return await payment_client.call(idempotency_key, total_price)

    success = await ctx.run_typed("payment", pay)

    if success:
        await ctx.run_typed(
            "send_success_email",
            email_client.notify_user_of_payment_success,
            user_id=order["user_id"],
        )
    else:
        await ctx.run_typed(
            "send_failure_email",
            email_client.notify_user_of_payment_failure,
            user_id=order["user_id"],
        )

    return success


# <end_checkout>
